cornice/schemas.py

```diff
@@ -1,180 +1,78 @@
+else:
+
+    class StrictMappingSchema(colander.MappingSchema):
+        @staticmethod
+        def schema_type():
+            return colander.Mapping(unknown='raise')
+
+    class CorniceSchema(colander.MappingSchema):
+        querystring = StrictMappingSchema()
+        headers = StrictMappingSchema()
+        body = StrictMappingSchema()
+        path = StrictMappingSchema()
+
+    def simple_cstruct_serialize(val):
+        """ cstruct is colander's internal use format which is a
+        series of nested dicts, lists, and tuples with only str values
+        as the "leaf" values.  If this method will try to recursively
+        go through a given value and convert all dict-like objects into
+        dicts, convert iterables that aren't dict-like into lists, and
+        convert everything else into a string.
         """
-        attributes = self.bind_attributes(request)
-
-        def _filter(attr):
-            if not hasattr(attr, "location"):
-                valid_location = 'body' in location
-            else:
-                valid_location = attr.location in to_list(location)
-            return valid_location and attr.required in to_list(required)
-
-        return list(filter(_filter, attributes))
-
-    def as_dict(self):
-        """returns a dict containing keys for the different attributes, and
-        for each of them, a dict containing information about them::
-
-            >>> schema.as_dict()  # NOQA
-            {'foo': {'type': 'string',
-                     'location': 'body',
-                     'description': 'yeah',
-                     'required': True},
-             'bar': {'type': 'string',
-                     'location': 'body',
-                     'description': 'yeah',
-                     'required': True}
-             # ...
-             }
-        """
-        attributes = self.bind_attributes()
-        schema = {}
-        for attr in attributes:
-            schema[attr.name] = {
-                'type': getattr(attr, 'type', attr.typ),
-                'name': attr.name,
-                'description': getattr(attr, 'description', ''),
-                'required': getattr(attr, 'required', False),
-            }
-
-        return schema
-
-    def unflatten(self, data):
-        return self.colander_schema.unflatten(data)
-
-    def flatten(self, data):
-        return self.colander_schema.flatten(data)
-
-    @classmethod
-    def from_colander(klass, colander_schema, **kwargs):
-        return CorniceSchema(colander_schema, **kwargs)
+        try:
+            # try dict-like interpretation
+            result = {}
+            for k in val.keys():
+                result[k] = simple_cstruct_serialize(val[k])
+            return result
+        except (TypeError, AttributeError):
+            try:
+                # try iterable interpretation
+                result = []
+                for k in val:
+                    result.append(simple_cstruct_serialize(k))
```

Comment:
yep.  it seems I didn't properly test `simple_cstruct_serialize()`.  It's because strings are iterable.  I'll patch this.